odl/trafos/non_uniform_fourier.py

```diff
@@ -9,123 +9,242 @@
-            Whether the normalization step should be skipped
+            Whether the samples normalization step should be skipped.
+            If ``True``, all ``samples`` are expected to lie between -0.5
+            (inclusive) and +0.5 (exclusive).
+        kwargs
+            Additional keyword arguments passed on to the ``NFFT`` class
+            constructor, except for its ``N`` and ``M`` parameters.
+
+        Notes
+        -----
+        - The ``pynfft`` backend library currently only supports double
+          precision arithmetic, thus all computations are carried out with
+          double precision irrespective of the ``dtype`` of ``domain`` and
+          ``range``.
         """
-        if not isinstance(space, DiscreteLp) or not space.is_uniform:
-            raise ValueError("`space` should be a uniform `DiscreteLp`")
         super(NonUniformFourierTransform, self).__init__(
-            space=space,
-            samples=samples,
-            domain=space,
-            range=cn(len(samples)),
-            skip_normalization=skip_normalization,
-        )
-
-    @property
-    def adjoint(self):
-        return NonUniformFourierTransformAdjoint(
-            shape=self.shape,
-            samples=self.samples,
-            skip_normalization=True,
+            domain, samples, range, **kwargs
         )
 
     def _call(self, x):
-        """Compute the direct non uniform FFT.
-
-        Parameters
-        ----------
-        x : `numpy.ndarray`
-            The data whose non uniform FFT you want to compute
-
-        Returns
-        -------
-        out_normalized : `numpy.ndarray`
-            Result of the transform
-        """
+        """Return ``self(x)``."""
         if not self._has_run:
-            self._normalize()
+            if not self._skip_normalization:
+                normalize_samples(
+                    self.samples, self.domain.cell_sides, out=self.samples
+                )
             self.nfft.precompute()
+            self.nfft.x = self.samples
             self._has_run = True
+
         self.nfft.f_hat = np.asarray(x)
         out = self.nfft.trafo()
-        # The normalization is inspired from
-        # https://github.com/CEA-COSMIC/pysap-mri/blob/master/mri/reconstruct/fourier.py#L123
-        out /= np.sqrt(self.nfft.M)
+        out *= self.domain.cell_volume / (2 * np.pi)
+        # TODO(kohr-h): phase factor needed
         return out
 
+    @property
+    def adjoint(self):
+        r"""Adjoint operator.
+
+        The adjoint is given by
+
+        .. math::
+            g_j =
+
+            = \sum_{k=0}^{M-1} \widehat{f}(\xi_k)\,
+            \mathrm{e}^{\mathrm{i} s j \cdot \xi_k},\quad
+            0 \leq j < N - 1.
+        """
+        return NonUniformFourierTransformAdjoint(
+            domain=self.range,
+            samples=self.samples,
+            range=self.domain,
+            nfft=self.nfft,
+            skip_normalization=self._has_run,
```

Comment:
I agree that the way I had done it before was wrong, but I think there is a problem here too.
Imagine a situation where we should skip normlization but the operator hasn't been run.
I think it should be:
`skip_normalization=self._has_run or self._skip_normalization`

(same for the adjoint)

Reply:
Hm, I'll have a look. Thx